avail-server/src/main/kotlin/com/avail/server/io/files/ServerFileWrapper.kt

```diff
@@ -0,0 +1,409 @@
+	 * next time the [file]'s local history is saved.
+	 */
+	private var tracedActionStackSavePointer: Int = 0
+
+	/**
+	 * Save the [tracedActionStack] to local history starting from the position
+	 * at [tracedActionStackSavePointer].
+	 */
+	fun conditionallySaveToLocalHistory ()
+	{
+		val currentSize = tracedActionStack.size
+		if (tracedActionStackSavePointer < currentSize)
+		{
+			val start = tracedActionStackSavePointer
+			for (i in start until  currentSize)
+			{
+				// TODO write this somewhere!
+				tracedActionStack[i]
+			}
+			tracedActionStackSavePointer = currentSize
+		}
+	}
+
+	/**
+	 * The [queue][ConcurrentLinkedQueue] of [FileAction]s that are waiting to
+	 * be executed for the contained [AvailServerFile].
+	 */
+	private val actionQueue = ConcurrentLinkedQueue<FileAction>()
+
+	/**
+	 * Is the [file] being available for [FileAction]a from the [actionQueue]?
+	 * `true` indicates it is; `false` otherwise.
+	 */
+	private val isReady = AtomicBoolean(true)
+
+	/**
+	 * Remove all the [FileAction]s from the [actionQueue] and perform them
+	 * in FIFO order if this [ServerFileWrapper] [is ready][isReady]
+	 * to update.
+	 */
+	private fun executeActions ()
+	{
+		if (isReady.getAndSet(false))
+		{
+			var action = actionQueue.poll()
+
+			// Check if an action is available and that the file hasn't been
+			// marked as being not ready before proceeding
+			while (action != null && isReady.get())
+			{
+				val tracedAction =
+					action.execute(file, System.currentTimeMillis())
+
+				if (undoStackDepth > 0)
+				{
+					// Must remove all TracedActions above the undoStackDepth
+					// pointer as they are no longer valid.
+					do
+					{
+						val ta = tracedActionStack.pop()
+						// TODO remove ta from the project cache?
+					}
+					while (--undoStackDepth > 0)
+				}
+				tracedActionStack.push(tracedAction)
+				action = actionQueue.poll()
+			}
+			// TODO is this safe
+			synchronized(actionQueue)
+			{
+				isReady.set(true)
+			}
+			// TODO Save file
+		}
+	}
+
+	/**
+	 * [Update][FileAction.execute] the wrapped [AvailServerFile] with the
+	 * provided [FileAction].
+	 *
+	 * @param fileAction
+	 *   The `FileAction` to perform.
+	 */
+	fun execute (fileAction: FileAction)
+	{
+		synchronized(actionQueue)
+		{
+			actionQueue.add(fileAction)
+		}
+		executeActions()
+	}
+
+	/**
+	 * The number of places from the top of the [tracedActionStack] that have
+	 * been undone.
+	 */
+	var undoStackDepth = 0
+
+	/**
+	 * [Undo][TracedAction.revert] the [FileAction] performed on the [file] from
+	 * the [TracedAction] that is [undoStackDepth] + 1 from the top of the
+	 * stack.
+	 */
+	fun undo ()
+	{
+		if (tracedActionStack.size < undoStackDepth + 1)
+		{
+			// We are at the bottom of the stack; there are no TracedActions
+			// eligible to be reverted.
+			return
+		}
+		tracedActionStack[tracedActionStack.size + ++undoStackDepth].revert(file)
```

Comment:
It's probably better to rewrite this to remove the early exit:

```kt
if (tracedActionStack.size >= undoStackDepth + 1)
{
	tracedActionStack[tracedActionStack.size + ++ undoStackDepth].revert(file)
}
```

Reply:
Done